### PR TITLE
LPS-186955 [FE] Unique Composite Key

### PR DESCRIPTION
**Pull request by igor-franca**
[Related](https://liferay.atlassian.net/browse/LPS-186955)

### CODE REVIEW

File: modules/apps/object/object-js-components-web/src/main/resources/META-INF/resources/components/Card.tsx [hunk 1]
```diff
@@ -36,6 +42,8 @@ export function Card({
 	viewMode,
 	...otherProps
 }: CardProps) {
+	const [showCardAlert, setShowCardAlert] = useState(true);
```
Comment (naraandrade) [inline modules/apps/object/object-js-components-web/src/main/resources/META-INF/resources/components/Card.tsx]: @igor-franca I think you can call this state just `showAlert`. We are not adding the prefix in the other properties like `noCardHeaderBorder` or `cardCustomHeader`.

File: modules/apps/object/object-web/src/main/resources/META-INF/resources/js/components/ObjectValidation/UniqueComposedKey.tsx [hunk 2]
```diff
@@ -32,7 +32,7 @@ export function UniqueComposedKey({
 		TBuilderScreenItem[]
 	>([]);
 
-	const filteredObjectFields = objectFields.filter(
+	const filteredCustomObjectFields = customObjectFields.filter(
 		(objectField) =>
```
Comment (naraandrade) [inline modules/apps/object/object-web/src/main/resources/META-INF/resources/js/components/ObjectValidation/UniqueComposedKey.tsx:88]: @igor-franca `customObjectField`

File: modules/apps/object/object-web/src/main/resources/META-INF/resources/js/components/ObjectValidation/UniqueComposedKey.tsx [hunk 3]
```diff
@@ -15,6 +15,22 @@ import React, {useEffect, useState} from 'react';
...
+	objectField: ObjectField;
+	objectValidationRuleSetting: ObjectValidationRuleSetting;
+	objectValidationRuleSettingNameMatches:
+		| 'keyObjectFieldExternalReferenceCode'
+		| 'outputObjectFieldExternalReferenceCode';
+}
+interface ModalSelectObjectFieldItemProps extends ObjectField {
+	checked: boolean;
+}
+interface MultipleSelectOptionProps {
```
Comment (naraandrade) [inline modules/apps/object/object-web/src/main/resources/META-INF/resources/js/components/ObjectValidation/UniqueComposedKey.tsx:31]: @igor-franca Please add blank line above.

File: modules/apps/object/object-web/src/main/resources/META-INF/resources/js/components/ObjectValidation/UniqueComposedKey.tsx [hunk 4]
```diff
@@ -39,10 +68,13 @@ export function UniqueComposedKey({
...
 		TBuilderScreenItem[]
 	>([]);
-	const [multipleSelectOptions, setMultipleSelectOptions] = useState<IItem[]>(
-		[]
-	);
-
+	const [
+		modalSelectObjectFieldsItems,
+		setModalSelectObjectFieldsItems,
+	] = useState<ModalSelectObjectFieldItemProps[]>([]);
```
Comment (naraandrade) [inline modules/apps/object/object-web/src/main/resources/META-INF/resources/js/components/ObjectValidation/UniqueComposedKey.tsx:79]: @igor-franca To follow the pattern of other states, please rename it to `ModalSelectObjectFieldItem`.

File: modules/apps/object/object-web/src/main/resources/META-INF/resources/js/components/ObjectValidation/UniqueComposedKey.tsx [hunk 5]
```diff
@@ -39,10 +68,13 @@ export function UniqueComposedKey({
...
+	const [multipleSelectOptions, setMultipleSelectOptions] = useState<
+		MultipleSelectOptionProps[]
```
Comment (naraandrade) [inline modules/apps/object/object-web/src/main/resources/META-INF/resources/js/components/ObjectValidation/UniqueComposedKey.tsx:81]: @igor-franca To follow the pattern of other states, please rename it to `MultipleSelectOption`.

File: modules/apps/object/object-js-components-web/src/main/resources/META-INF/resources/components/Select/MultipleSelect.tsx [hunk 6]
```diff
@@ -73,6 +73,7 @@ export function MultipleSelect<
 		<BaseSelect
 			{...restProps}
 			dropdownActive={dropdownActive}
+			required={required}
```
Comment (naraandrade) [inline modules/apps/object/object-js-components-web/src/main/resources/META-INF/resources/components/Select/MultipleSelect.tsx:76]: @igor-franca The commit message doesn't seem to describe the changes here.

**Review by naraandrade — CHANGES_REQUESTED** [review]

File: modules/apps/portal-language/portal-language-lang/src/main/resources/content/Language.properties [hunk 7]
```diff
@@ -699,6 +700,7 @@ add-a-high-precision-decimal-number-without-rounding=Add a high precision decima
 add-a-label=Add a label.
 add-a-long-integer-up-to-16-digits=Add a long integer up to 16 digits.
 add-a-message-here=Add a message here.
+add-a-minimum-of-two-fields-to-create-compose-unique-keys=Add a minimum of two fields to create Compose Unique Keys.
```
[the remark this answers is not in the document]
Comment (igor-franca) [reply modules/apps/portal-language/portal-language-lang/src/main/resources/content/Language.properties:703]: I will check, thanks!

**Comment by igor-franca** [comment]
Thank you for this very complete review, Nara! I just completed your requests.

**Review by naraandrade — CHANGES_REQUESTED** [review]

File: modules/apps/object/object-web/src/main/resources/META-INF/resources/js/components/ObjectAction/PredefinedValuesTable.tsx [hunk 8]
```diff
@@ -253,7 +253,7 @@ export default function PredefinedValuesTable({
 		const handleAddFields = () => {
```
[the remark this answers is not in the document]
Comment (igor-franca) [reply modules/apps/object/object-web/src/main/resources/META-INF/resources/js/components/ObjectAction/PredefinedValuesTable.tsx]: Thanks for review @naraandrade, I performed this and other changes related object prefix in this file 👍🏾

File: modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectValidationRuleLocalServiceImpl.java [hunk 9]
```diff
@@ -226,6 +226,16 @@ public List<ObjectValidationRule> getObjectValidationRules(
 				objectDefinitionId, active));
 	}
 
+	@Override
+	public List<ObjectValidationRule>
+		getObjectValidationRulesByObjectValidationRuleEngine(
+			String engine, long objectDefinitionId) {
```
Comment (pedro-oliveira446) [inline modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectValidationRuleLocalServiceImpl.java:232]: Wouldn't the order here be the same as in the service.xml file?

**Review by pedro-oliveira446 — CHANGES_REQUESTED** [review]

File: modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectValidationRuleLocalServiceImpl.java (hunk 9 above)
Comment (igor-franca) [reply modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectValidationRuleLocalServiceImpl.java]: Good catch, thanks @pedro-oliveira446

**Review by pedro-oliveira446 — APPROVED** [review]

File: modules/apps/object/object-web/src/main/resources/META-INF/resources/js/components/ObjectValidation/AddObjectValidation.tsx [hunk 10]
```diff
@@ -25,7 +25,7 @@ interface ModalAddObjectValidationProps extends AddObjectValidationProps {
 
 interface AddObjectValidationProps {
 	apiURL: string;
-	objectValidationRuleEngines: ObjectValidationType[];
+	objectValidationRuleEngines: LabelNameObject[];
```
[the remark this answers is not in the document]
Comment (igor-franca) [reply modules/apps/object/object-web/src/main/resources/META-INF/resources/js/components/ObjectValidation/AddObjectValidation.tsx:28]: Hey, Nara. I removed [here](https://github.com/liferay-objects/liferay-portal/pull/2688/commits/636a2ab569f1c343e1e6fbf86596db81d3f17c9e#diff-63608e745a4c720211693896469a7bf97d034829944b98edab0d2ed29a092d7cL445-L449), but I believe its returned inside some any conflict resolution. Good catch!

**Comment by igor-franca** [comment]
ci:test:sf

**Review by naraandrade — APPROVED** [review]

**Comment by igor-franca** [comment]
The next push will be to make the code compatible with the latest BE modifications:

File: modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectValidationRuleLocalServiceImpl.java [hunk 11]
```diff
@@ -226,6 +226,16 @@ public List<ObjectValidationRule> getObjectValidationRules(
 				objectDefinitionId, active));
 	}
 
+	@Override
+	public List<ObjectValidationRule>
+		getObjectValidationRulesByObjectValidationRuleEngine(
```
Comment (carolmariaabb) [inline modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectValidationRuleLocalServiceImpl.java:231]: @igor-franca please rename it to `getObjectValidationRules` if possible. If not, rename it to `getObjectValidationRulesByODI_E`.

File: modules/apps/object/object-web/src/main/java/com/liferay/object/web/internal/object/definitions/portlet/action/GetObjectFieldDeleteInfoMVCResourceCommand.java [hunk 12]
```diff
@@ -50,34 +56,82 @@ protected void doServeResource(
...
+				"deleteObjectFieldObjectValidationRuleSetting",
+				() -> _shouldDeleteObjectFieldObjectValidationRuleSetting(
+					objectDefinition, objectField)
+			).put(
 				"showDeletionModal",
-				() -> {
-					if (!objectDefinition.isApproved()) {
-						return false;
-					}
+				() -> _shouldShowDeletionModal(objectDefinition, objectField)
```
Comment (carolmariaabb) [inline modules/apps/object/object-web/src/main/java/com/liferay/object/web/internal/object/definitions/portlet/action/GetObjectFieldDeleteInfoMVCResourceCommand.java:68]: @igor-franca please inline as this method is only used once:

```
.put(
	"showDeletionModal",
	() -> {
		if (objectDefinition.isApproved() &&
			_shouldDeleteLastPublishedObjectDefinitionObjectField(
				objectDefinition, objectField) &&
			_shouldDeleteObjectFieldObjectValidationRuleSetting(
				objectDefinition, objectField)) {

			return true;
		}

		return false;
	}
)
```

File: modules/apps/object/object-web/src/main/java/com/liferay/object/web/internal/object/definitions/portlet/action/GetObjectFieldDeleteInfoMVCResourceCommand.java [hunk 13]
```diff
@@ -50,34 +56,82 @@ protected void doServeResource(
...
+	private boolean _shouldDeleteObjectFieldObjectValidationRuleSetting(
+		ObjectDefinition objectDefinition, ObjectField objectField) {
 
-					int customObjectFieldsCount =
-						_objectFieldLocalService.getObjectFieldsCount(
-							objectField.getObjectDefinitionId(), false);
+		for (ObjectValidationRule objectValidationRule :
+				_objectValidationRuleLocalService.
+					getObjectValidationRulesByObjectValidationRuleEngine(
+						objectDefinition.getObjectDefinitionId(),
```
Comment (carolmariaabb) [inline modules/apps/object/object-web/src/main/java/com/liferay/object/web/internal/object/definitions/portlet/action/GetObjectFieldDeleteInfoMVCResourceCommand.java:96]: @igor-franca we can use the `getObjectDefinitionId()` from the `objectField` to simplify, i.e.:

```
private boolean _shouldDeleteObjectFieldObjectValidationRuleSetting(
	ObjectField objectField) {

	for (ObjectValidationRule objectValidationRule :
			_objectValidationRuleLocalService.getObjectValidationRules(
				objectField.getObjectDefinitionId(),
```

File: modules/apps/object/object-web/src/main/java/com/liferay/object/web/internal/object/definitions/portlet/action/GetObjectFieldDeleteInfoMVCResourceCommand.java [hunk 14]
```diff
@@ -50,34 +56,82 @@ protected void doServeResource(
...
+					objectValidationRule.getObjectValidationRuleSettings()) {
+
+				if (objectValidationRuleSetting.compareName(
+						ObjectValidationRuleSettingConstants.
+							NAME_COMPOSITE_KEY_OBJECT_FIELD_ID)) {
+
+					long objectValidationRuleSettingId = GetterUtil.getLong(
+						objectValidationRuleSetting.getValue());
+
+					if (objectValidationRuleSettingId ==
```
Comment (carolmariaabb) [inline modules/apps/object/object-web/src/main/java/com/liferay/object/web/internal/object/definitions/portlet/action/GetObjectFieldDeleteInfoMVCResourceCommand.java:110]: @igor-franca please rename it to `objectFieldId`

```
long objectFieldId = GetterUtil.getLong(
	objectValidationRuleSetting.getValue());

if (objectFieldId == objectField.getObjectFieldId()) {
```

File: modules/apps/object/object-web/src/main/java/com/liferay/object/web/internal/object/definitions/portlet/action/GetObjectFieldDeleteInfoMVCResourceCommand.java [hunk 15]
```diff
@@ -50,34 +56,82 @@ protected void doServeResource(
...
+						objectDefinition.getObjectDefinitionId(),
+						ObjectValidationRuleConstants.
+							ENGINE_TYPE_COMPOSITE_KEY)) {
+
+			for (ObjectValidationRuleSetting objectValidationRuleSetting :
+					objectValidationRule.getObjectValidationRuleSettings()) {
+
+				if (objectValidationRuleSetting.compareName(
+						ObjectValidationRuleSettingConstants.
+							NAME_COMPOSITE_KEY_OBJECT_FIELD_ID)) {
```
Comment (carolmariaabb) [inline modules/apps/object/object-web/src/main/java/com/liferay/object/web/internal/object/definitions/portlet/action/GetObjectFieldDeleteInfoMVCResourceCommand.java:105]: @igor-franca we can buy some space here:

```
if (!objectValidationRuleSetting.compareName(
	  ObjectValidationRuleSettingConstants.
		NAME_COMPOSITE_KEY_OBJECT_FIELD_ID)) {

	continue;
}

long objectFieldId = GetterUtil.getLong(
	objectValidationRuleSetting.getValue());

if (objectFieldId == objectField.getObjectFieldId()) {
	return false;
}
```

**Review by carolmariaabb — CHANGES_REQUESTED** [review]

**Review by carolmariaabb — APPROVED** [review]
LGTM!

**Comment by carolmariaabb** [comment]
ci:test:sf

**Comment by igor-franca** [comment]
ci:test:relevant

**Comment by igor-franca** [comment]
ci:test:object

**Review by naraandrade — APPROVED** [review]

**Comment by mateussgn** [comment]
ci:forward:force

**Review by carolmariaabb — APPROVED** [review]

**Comment by carolmariaabb** [comment]
ci:forward

**Comment by igor-franca** [comment]
ci:forward:force

**Comment by mateussgn** [comment]
ci:forward:force

**Pull request closed without merging**